Mathlib/CategoryTheory/ConcreteCategory/Basic.lean

```diff
@@ -225,4 +238,187 @@ lemma NatTrans.naturality_apply {C D : Type*} [Category C] [Category D] [HasForg
+
+namespace ConcreteCategory
+
+attribute [simp] id_apply comp_apply
+
+/-- We can apply morphisms of concrete categories by first casting them down
+to the base functions.
+-/
+instance {X Y : C} : CoeFun (X ⟶ Y) (fun _ ↦ ToType X → ToType Y) where
+  coe f := hom f
+
+/--
+`ConcreteCategory.hom` bundled as an `Equiv`.
+-/
+def homEquiv {X Y : C} : (X ⟶ Y) ≃ ToHom X Y where
+  toFun := hom
+  invFun := ofHom
+  left_inv := ofHom_hom
+  right_inv := hom_ofHom
+
+lemma hom_bijective {X Y : C} : Function.Bijective (hom : (X ⟶ Y) → ToHom X Y) :=
+  homEquiv.bijective
+
+lemma hom_injective {X Y : C} : Function.Injective (hom : (X ⟶ Y) → ToHom X Y) :=
+  hom_bijective.injective
+
+/-- In any concrete category, we can test equality of morphisms by pointwise evaluations. -/
+@[ext] lemma ext {X Y : C} {f g : X ⟶ Y} (h : hom f = hom g) : f = g :=
+  hom_injective h
+
+lemma coe_ext {X Y : C} {f g : X ⟶ Y} (h : ⇑(hom f) = ⇑(hom g)) : f = g :=
+  ext (DFunLike.coe_injective h)
+
+lemma ext_apply {X Y : C} {f g : X ⟶ Y} (h : ∀ x, f x = g x) : f = g :=
+  ext (DFunLike.ext _ _ h)
+
+/-- A concrete category comes with a forgetful functor to `Type`.
+
+Warning: because of the way that `ConcreteCategory` and `HasForget` are set up, we can't make
+`forget Type` reducibly defeq to the identity functor. -/
+instance toHasForget : HasForget C where
+  forget.obj := ToType
+  forget.map f := ⇑(hom f)
+  forget_faithful.map_injective h := coe_ext h
+
+end ConcreteCategory
+
+theorem forget_obj (X : C) : (forget C).obj X = ToType X := rfl
+
+@[simp]
+theorem ConcreteCategory.forget_map_eq_coe {X Y : C} (f : X ⟶ Y) : (forget C).map f = f := rfl
+
+/-- Analogue of `congr_fun h x`,
+when `h : f = g` is an equality between morphisms in a concrete category.
+-/
+protected theorem congr_fun {X Y : C} {f g : X ⟶ Y} (h : f = g) (x : ToType X) : f x = g x :=
+  congrFun (congrArg (fun k : X ⟶ Y => (k : ToType X → ToType Y)) h) x
+
+/-- Analogue of `congr_arg f h`,
+when `h : x = x'` is an equality between elements of objects in a concrete category.
+-/
+protected theorem congr_arg {X Y : C} (f : X ⟶ Y) {x x' : ToType X} (h : x = x') : f x = f x' :=
+  congrArg (f : ToType X → ToType Y) h
+
+theorem hom_id {X : C} : (𝟙 X : ToType X → ToType X) = id :=
+  (forget _).map_id X
+
+theorem hom_comp {X Y Z : C} (f : X ⟶ Y) (g : Y ⟶ Z) : (f ≫ g : ToType X → ToType Z) = g ∘ f :=
+  (forget _).map_comp f g
+
+section
+
+variable (C)
+
+/-- Build a coercion to functions out of `HasForget`.
+
+The intended usecase is to provide a `FunLike` instance in `HasForget.toConcreteCategory`.
+See that definition for the considerations in making this an instance.
+
+See note [reducible non-instances].
+-/
+abbrev HasForget.toFunLike [HasForget C] (X Y : C) :
+    FunLike (X ⟶ Y) ((forget C).obj X) ((forget C).obj Y) where
+  coe := (forget C).map
+  coe_injective' _ _ h := Functor.Faithful.map_injective h
+
+attribute [local instance] HasForget.toFunLike
```

Comment:
Why are these set? Just preparation for the incremental swap?

Reply:
The `local instance` attribute? That's needed to make the instances available for the declarations below (like `forget_eq_ConcreteCategory_hom`).

Reply:
Do you want to remove the previous lines enabling them?

Reply:
Looks like those can indeed be removed, thanks!